pkg/ruler/manager.go

```diff
@@ -105,9 +105,17 @@ func (r *DefaultMultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGrou
 		RemoveFederatedRuleGroups(ruleGroups)
 	}
 
+	// Sync the rules to disk and then update the user's Prometheus Rules Manager.
+	// Since users are different, we can sync rules in parallel.
+	var wg sync.WaitGroup
 	for userID, ruleGroup := range ruleGroups {
-		r.syncRulesToManager(ctx, userID, ruleGroup)
+		wg.Add(1)
+		go func(id string, rg rulespb.RuleGroupList) {
+			defer wg.Done()
+			r.syncRulesToManager(ctx, id, rg)
```

Comment:
Can you make sure this function is concurrent safe, please?

Reply:
I wanted to test it here by adding more user in this [Test](https://github.com/grafana/mimir/pull/4258/files#diff-44e204d29de314b23cb0c64cc703e31ba964ccc6bdc079d7b10385736f3c3019R48), I add [this test](https://github.com/grafana/mimir/pull/4258/files#diff-44e204d29de314b23cb0c64cc703e31ba964ccc6bdc079d7b10385736f3c3019R125) for explicitly test syncRulesToManager.